CGHub download

jfeala opened the pull request:
Hi Uri

Here is a bit of code to follow-up to our emails. This is **untested** and not ready for merge, but I wanted to get your feedback before continuing.

CGHub is hosting > 1200 public BAM files from the Cancer Cell Line Encyclopedia that are available only through their GeneTorrent download client. TCGA BAM files, while not available publicly, can be downloaded using the same framework for anyone with an authorized key file.

The dependencies for installing GeneTorrent are tricky on CentOS so that part would require the most attention. The other changes are straightforward.

I invented a url of the form `cghub://<analysis_id>/<filename>` to distinguish these downloads but we should think about whether there is a better way.

Jake

Commit:
cghub download via GeneTorrent

Commit:
CCLE BAM files registry

jfeala commented:
... and a little background:   https://cghub.ucsc.edu/datasets/ccle.html

fnothaft commented:
+1

tomwhite commented on registry/ccle-wgs.json:
```diff
@@ -0,0 +1,31 @@
+{
+    "name": "ccle-wgs",
+    "description": "Cancer Cell Line Encyclopedia whole genome sequencing",
+    "target": "ccle/wgs",
```
Nit: We've updated the format recently, so that `target` is no longer used (see https://github.com/bigdatagenomics/eggo/blob/master/docs/spec.md), so this line can be omitted. (I still need to update the other files in the registry).

tomwhite commented:
This looks good to me. It might be helpful to add a file in test/registry to make it easy to try out on a single  file from CGHub.

Commit:
added test registry for cghub files

Commit:
removed "target" field from registries

Commit:
added TCGA WGS benchmark registry

jfeala commented:
I updated the registries according to your suggestions, and added the other public dataset available on CGHub, a benchmarking dataset published for the purpose of testing mutation callers. Unfortunately the smallest file from both studies is a ~5Gb RNA-seq BAM, so I used that one as the test registry.

Still testing the ETL code, not ready for merge yet

Commit:
cleanup and debug

jfeala commented:
Ready for merge. GeneTorrent installation and CGHub download functions are tested. I couldn't get the full luigi DAG working, but it seems like that's a work in progress.

laserson commented on eggo/dag.py:
```diff
@@ -69,6 +69,40 @@ class ToastConfig(Config):
+        cghub_key = os.environ.get('CGHUB_KEY') or CGHUB_PUBLIC_KEY
+
+    # 2. Parse url for analysis ID and filename
+    analysis_id, filename = url.lstrip('cghub://').split('/')
```
Does the split ever produce more than two objects?

laserson replied:
`analysis_id` is a CGHub concept?

jfeala replied:
yes, the CGHub metadata store centers around the `analysis_id`. It refers to a single downloadable object (which may contain multiple files). I made up a cghub url to fit the existing registry structure, so it is easy to change. Right now I am creating them to only contain the analysis ID and a single filename of interest (BAM file, generally), so this would always split into 2 objects. 

However, the CGHub REST API returns a JSON with lots of metadata for a given analysis_id. One option would be to store this full JSON in the registry, though it would be long and cluttered, not as easily human-readable. Or we could just use the analysis ID and have the code call the API to get the filename, filesize, and other metadata if necessary.

laserson commented on eggo/dag.py:
```diff
@@ -69,6 +69,40 @@ class ToastConfig(Config):
+                         analysis_id=analysis_id), shell=True)
+    p.wait()
+
+    return os.path.join(tmp_dir, analysis_id)
```
I'd prefer to get rid of this `return` statement to be consistent with `_http_download`.

jfeala replied:
sure. this is to prevent nested folders in the eggo bucket, since gtdownload creates a folder named by analysis_id and stores the files inside it. I'll have it copy them out into tmp_dir instead

laserson commented on eggo/dag.py:
```diff
@@ -78,10 +112,12 @@ def _dnload_to_local_upload_to_s3(source, destination, compression):
+            _http_download(source, tmp_dir)
+        elif source.startswith('cghub'):
+            tmp_dir = _cghub_download(source, tmp_dir)
+        else:
```
What about `ftp`?

laserson commented on eggo/dag.py:
```diff
@@ -96,13 +132,13 @@ def _dnload_to_local_upload_to_s3(source, destination, compression):
         # 3. upload to tmp S3 location
         tmp_s3_path = os.path.join(EGGO_S3_TMP_URL, random_id())
-        upload_cmd = 'pushd {tmp_dir} && aws s3 cp ./* {s3_path} && popd'
+        upload_cmd = 'pushd {tmp_dir} && aws s3 cp . {s3_path} --recursive && popd'
```
What's the rationale behind this change?  At the moment, I think that additional levels of nested directories will break our URL scheme.  But perhaps this is something we need to support?  cc @tomwhite

jfeala replied:
globs weren't working for me with `aws s3 cp` when testing locally with the latest version of awscli, but `--recursive` worked fine. Didn't know you were purposely restricting nested folders. What version are you using?

laserson replied:
hmm, I'm pretty sure the latest version gets installed on the cluster.  This change is fine with me as long as the result is the same.  But haven't had a problem yet.

laserson commented on eggo/dag.py:
```diff
@@ -96,13 +132,13 @@ def _dnload_to_local_upload_to_s3(source, destination, compression):
 
         # 4. rename to final target location
-        rename_cmd = 'aws s3 mv {tmp_path} {final_path}'
+        rename_cmd = 'aws s3 mv {tmp_path} {final_path} --recursive'
```
same comment

Commit:
Standardize download functions

Standardize behavior of cghub, ftp, and http download functions
- flatten the cghub download directory and omit return value
- rename http_download to curl download to capture ftp use case
Also add “editions” field to test-cghub.json

Commit:
Merge upstream master

Commit:
Bugfix merge conflict

jfeala commented:
Ok, fixed it according to your suggestions (although the awscli globbing issue is now moot after recent updates to master). Let me know if you prefer a rebase
